[JSC] Use simple Vector for BufferedStructures

Constellation opened the pull request:
#### b8454d3a096356e8431c748f73417527eefca678
<pre>
[JSC] Use simple Vector for BufferedStructures
<a href="https://bugs.webkit.org/show_bug.cgi?id=273863">https://bugs.webkit.org/show_bug.cgi?id=273863</a>
<a href="https://rdar.apple.com/127713950">rdar://127713950</a>

Reviewed by Keith Miller.

This HashMap size is relatively small (since we have upper limit for IC feedback).
We should make it Vector since HashMap for this size is costly.

* Source/JavaScriptCore/bytecode/StructureStubInfo.cpp:
(JSC::StructureStubInfo::visitAggregateImpl):
(JSC::StructureStubInfo::visitWeakReferences):
* Source/JavaScriptCore/bytecode/StructureStubInfo.h:
(JSC::StructureStubInfo::considerRepatchingCacheImpl):
(JSC::StructureStubInfo::clearBufferedStructures):

Canonical link: <a href="https://commits.webkit.org/278530@main">https://commits.webkit.org/278530@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/89f28e15ac2bd8efc0567d5c7a97ed3c899dc417

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/50842 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/30139 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/3160 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/54101 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/1533 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/36415 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/1184 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/41425 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/52941 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/27787 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/43799 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/22553 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/25130 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/1055 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/9283 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/44178 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/47112 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/1117 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/55695 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/50345 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/25944 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/1004 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/48827 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/27201 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/43874 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/47909 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/28069 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/57820 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/7370 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/26933 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/11910 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

hyjorc1 commented on Source/JavaScriptCore/bytecode/StructureStubInfo.cpp:
```diff
@@ -336,12 +336,17 @@ void StructureStubInfo::reset(const ConcurrentJSLockerBase& locker, CodeBlock* c
-        for (auto& bufferedStructure : m_bufferedStructures)
-            bufferedStructure.byValId().visitAggregate(visitor);
-    }
-    m_identifier.visitAggregate(visitor);
```
Why previously we don't check the boolean context of `m_identifier`?

hyjorc1 replied:
Oh, I see. `visitAggregate`'s implementation does have a precondition including the `bool` operation.
```
template<typename Visitor>
inline void CacheableIdentifier::visitAggregate(Visitor& visitor) const
{
    if (m_bits && isCell())
        visitor.appendUnbarriered(cell());
}

explicit operator bool() const { return m_bits; }
```

Achierius commented on Source/JavaScriptCore/bytecode/StructureStubInfo.h:
```diff
@@ -431,7 +462,7 @@ class StructureStubInfo {
     // structure again during this buffering countdown, we will create an AccessCase object for it.
     // That's not so bad - we'll get rid of the redundant ones once we regenerate.
-    HashSet<BufferedStructure, BufferedStructure::Hash, BufferedStructure::KeyTraits> m_bufferedStructures WTF_GUARDED_BY_LOCK(m_bufferedStructuresLock);
+    std::variant<std::monostate, Vector<StructureID>, Vector<std::tuple<StructureID, CacheableIdentifier>>> m_bufferedStructures WTF_GUARDED_BY_LOCK(m_bufferedStructuresLock);
```
Why is the std::monostate case necessary? At first glance it feels like a size-0 vector would serve the same purpose & be just as default-constructible

Constellation replied:
`StructureStubInfo` is initialized before it gets information about AccessType, so at first, it does not know which one is required.

Achierius reviewed (commented):
If we know that the number of elements stored is usually very small, could we use something like llvm::SmallVector instead to avoid some allocations?

kmiller68 commented on Source/JavaScriptCore/bytecode/StructureStubInfo.h:
```diff
@@ -297,9 +297,33 @@ class StructureStubInfo {
+                        isNewlyAdded = true;
+                    },
+                    [&](Vector<std::tuple<StructureID, CacheableIdentifier>>& structures) {
+                        for (auto& [bufferedStructureID, bufferedCacheableIdentifier] : structures) {
```
Probably worth an `ASSERT(!m_identifier)` here.

Constellation replied:
Done.

kmiller68 commented on Source/JavaScriptCore/bytecode/StructureStubInfo.h:
```diff
@@ -431,7 +462,7 @@ class StructureStubInfo {
     // Note that it's always safe to clear this. If we clear it prematurely, then if we see the same
     // structure again during this buffering countdown, we will create an AccessCase object for it.
     // That's not so bad - we'll get rid of the redundant ones once we regenerate.
-    HashSet<BufferedStructure, BufferedStructure::Hash, BufferedStructure::KeyTraits> m_bufferedStructures WTF_GUARDED_BY_LOCK(m_bufferedStructuresLock);
```
Can you delete `BufferedStructure`. I don't think it's used anymore?

Constellation replied:
Done.

kmiller68 reviewed (approved):
r=me with comments.

Constellation committed:
[JSC] Use simple Vector for BufferedStructures
https://bugs.webkit.org/show_bug.cgi?id=273863
rdar://127713950

Reviewed by Keith Miller.

This HashMap size is relatively small (since we have upper limit for IC feedback).
We should make it Vector since HashMap for this size is costly.

* Source/JavaScriptCore/bytecode/StructureStubInfo.cpp:
(JSC::StructureStubInfo::visitAggregateImpl):
(JSC::StructureStubInfo::visitWeakReferences):
* Source/JavaScriptCore/bytecode/StructureStubInfo.h:
(JSC::StructureStubInfo::considerRepatchingCacheImpl):
(JSC::StructureStubInfo::clearBufferedStructures):

Canonical link: https://commits.webkit.org/278530@main
